reducers/plugins/index.js

```diff
@@ -136,10 +184,20 @@ function components(state = {}, action) {
     case ActionTypes.RECEIVED_PLUGIN_COMPONENT: {
```

Comment:
@scottleedavis How about if we do something like:
```
import remove from 'lodash/remove';
...
...


function hasMenuId(menu, menuId) {
    if (!menu.subMenu) {
        return false;
    }

    if (menu.id === menuId) {
        return true;
    }
    for (const subMenu of menu.subMenu) {
        // Recursively check if subMenu contains menuId.
        if (hasMenuId(subMenu, menuId)) {
            return true;
        }
    }
    return false;
}

function buildMenu(rootMenu, data) {
    // Recursively build the full menu tree.
    const subMenu = rootMenu.subMenu.map((m) => buildMenu(m, data));
    if (rootMenu.id === data.parentMenuId) {
        subMenu.push(data);
    }

    return {
        ...rootMenu,
        subMenu,
    };
}

...
...


function components(state = {}, action) {
    switch (action.type) {
    case ActionTypes.RECEIVED_PLUGIN_COMPONENT: {
        if (action.name && action.data) {
            const nextState = {...state};
            const currentArray = [...(nextState[action.name] || [])];
            let actionData = action.data;
            if (action.name === 'PostDropdownMenu' && actionData.parentMenuId) {
                // Remove the menu from currentArray to rebuild it later.
                const menu = remove(currentArray, (c) => hasMenuId(c, actionData.parentMenuId) && c.pluginId === actionData.pluginId);

                // Request is for an unknown menuId, return original state.
                if (!menu[0]) {
                    return state;
                }
                actionData = buildMenu(menu[0], actionData);
            }
            const nextArray = [...currentArray, actionData];
            nextArray.sort(sortComponents);
            nextState[action.name] = nextArray;
            return nextState;
        }
        return state;
    }
    case ActionTypes.REMOVED_PLUGIN_COMPONENT:
        return removePluginComponent(state, action);
    case ActionTypes.RECEIVED_WEBAPP_PLUGIN:
    case ActionTypes.REMOVED_WEBAPP_PLUGIN:
        return removePluginComponents(state, action);
    default:
        return state;
    }
}
```

Reply:
Very clean, nice work @ali-farooq0 !  :) I didn't realize we had lodash at our convenience.  Originally, I didn't like the offset into found/removed menus instead of a 'find' (e.g. menu[0]), however the clean approach has grown on me.  

Verified it works with various submenus from the same plugin combinations.
<img width="964" alt="Screen Shot 2019-10-07 at 7 38 07 PM" src="https://user-images.githubusercontent.com/2914443/66363511-621e3580-e93b-11e9-8204-0c7b361ccd09.png">

Reply:
Sweet!! Thanks for verifying @scottleedavis! I added a minor tweak on how `nextArray` is created. Didn't want to trouble you with the minor issue that came out of my change :(, But should be good now. Mind taking another look to make sure all is well?

Reply:
Passed all my usage tests, again nice work!